builder/dockerfile/evaluator.go

```diff
@@ -25,178 +25,182 @@ import (
+	return c, ok
+}
 
-	args := []string{}
-	ast := node
-	if cmd == command.Onbuild {
-		var err error
-		ast, args, err = handleOnBuildNode(node, msg)
-		if err != nil {
-			return nil, err
-		}
+func (r *stagesBuildResults) validateIndex(i int) error {
+	if i == len(r.flat) {
+		return errors.New("refers to current build stage")
 	}
+	if i < 0 || i > len(r.flat) {
+		return errors.New("index out of bounds")
+	}
+	return nil
+}
 
-	runConfigEnv := options.state.runConfig.Env
-	envs := append(runConfigEnv, b.buildArgs.FilterAllowed(runConfigEnv)...)
-	processFunc := createProcessWordFunc(options.shlex, cmd, envs)
-	words, err := getDispatchArgsFromNode(ast, processFunc, msg)
+func (r *stagesBuildResults) get(nameOrIndex string) (*container.Config, error) {
+	if c, ok := r.getByName(nameOrIndex); ok {
+		return c, nil
+	}
+	ix, err := strconv.ParseInt(nameOrIndex, 10, 0)
 	if err != nil {
-		buildsFailed.WithValues(metricsErrorProcessingCommandsError).Inc()
+		return nil, nil
+	}
+	if err := r.validateIndex(int(ix)); err != nil {
 		return nil, err
 	}
-	args = append(args, words...)
+	return r.flat[ix], nil
+}
 
-	fmt.Fprintln(b.Stdout, msg.String())
+func (r *stagesBuildResults) checkStageNameAvailable(name string) error {
+	if name != "" {
+		if _, ok := r.getByName(name); ok {
+			return errors.Errorf("%s stage name already used", name)
+		}
+	}
+	return nil
+}
 
-	f, ok := evaluateTable[cmd]
-	if !ok {
-		buildsFailed.WithValues(metricsUnknownInstructionError).Inc()
-		return nil, fmt.Errorf("unknown instruction: %s", upperCasedCmd)
+func (r *stagesBuildResults) commitStage(name string, config *container.Config) error {
+	if name != "" {
+		if _, ok := r.getByName(name); ok {
+			return errors.Errorf("%s stage name already used", name)
+		}
+		r.indexed[strings.ToLower(name)] = config
 	}
-	options.state.updateRunConfig()
-	err = f(newDispatchRequestFromOptions(options, b, args))
-	return options.state, err
+	r.flat = append(r.flat, config)
+	return nil
 }
 
-type dispatchOptions struct {
+type dispatchRequest struct {
 	state   *dispatchState
-	stepMsg string
-	node    *parser.Node
 	shlex   *ShellLex
+	builder *Builder
 	source  builder.Source
+	stages  *stagesBuildResults
 }
 
-// dispatchState is a data object which is modified by dispatchers
-type dispatchState struct {
-	runConfig  *container.Config
-	maintainer string
-	cmdSet     bool
-	imageID    string
-	baseImage  builder.Image
-	stageName  string
+func newDispatchRequest(builder *Builder, escapeToken rune, source builder.Source, buildArgs *buildArgs, stages *stagesBuildResults) *dispatchRequest {
+	return &dispatchRequest{
+		state:   newDispatchState(buildArgs),
+		shlex:   NewShellLex(escapeToken),
+		builder: builder,
+		source:  source,
+		stages:  stages,
+	}
 }
 
-func newDispatchState() *dispatchState {
-	return &dispatchState{runConfig: &container.Config{}}
+func (d *dispatchRequest) commitStage() error {
+	return d.stages.commitStage(d.state.stageName, d.state.runConfig)
```

Comment:
`commitStage()` sounds like a very strange behaviour for a `dispatchRequest()`. This probably belongs somewhere else.